File: 2/ucs.pl
uniform_cost_search(Start, Goal, Edges, Path, Cost) :-
    empty_queue(EmptyQueue),
    enqueue([(Start, [])], EmptyQueue, Queue),
    ucs_loop(Goal, Edges, Queue, Path, Cost).

ucs_loop(Goal, _, Queue, Path, Cost) :-
    dequeue([(Current, PathSoFar)|_], Queue, _),
    Current = Goal,
    reverse(PathSoFar, Path),
    cost(Path, Cost).
ucs_loop(Goal, Edges, Queue, Path, Cost) :-
    dequeue([(Current, PathSoFar)|Rest], Queue, _),
    \+ Current = Goal,
    expand(Current, PathSoFar, Edges, NewPaths),
    add_paths_to_queue(NewPaths, Queue, NewQueue),
    ucs_loop(Goal, Edges, NewQueue, Path, Cost).

expand(Node, PathSoFar, Edges, NewPaths) :-
    findall((Next, [Next|PathSoFar]),
            (member((Node, Next, Cost), Edges)),
            NextPaths),
    maplist(add_cost_to_path(Cost), NextPaths, NewPaths).

add_cost_to_path(Cost, (Node, Path), (Node, NewPath)) :-
    cost(Path, OldCost),
    NewCost is OldCost + Cost,
    append(Path, [Node], NewPath),
    cost(NewPath, NewCost).

cost(Path, Cost) :-
    maplist(get_cost, Path, Costs),
    sum_list(Costs, Cost).

get_cost(Node, Cost) :-
    member((_, Node, Cost), Edges).

add_paths_to_queue([], Queue, Queue).
add_paths_to_queue([Path|Paths], Queue, NewQueue) :-
    enqueue([Path], Queue, TempQueue),
    add_paths_to_queue(Paths, TempQueue, NewQueue).

File: 2/ucs.py
from queue import PriorityQueue
g1={
    'A':{'B':4,'C':7},
    'B':{'D':3,'E':6},
    'C':{'F':2},
    'D':{'G':4,'H':2},
    'E':{},
    'F':{'I':4,'J':7},
    'G':{},
    'H':{'L':6},
    'I':{},
    'J':{'K':8},
    'K':{},
    'L':{},
}
parent={}
q=PriorityQueue()
visited={}
distance={}
for i in g1:
    visited[i]=False
    distance[i]=-1
distance['A']=0
curr='A'
q.put((0,'A'))
visited['A']=True
path=[]
while q.qsize()!=0:
    curr=q.get()[1]
    if(curr=='G'):
        print(distance['G'])
        j='G'
        path.append('G')
        while(j!='A'):
            j=parent[j]
            path.append(j)
        path.reverse()
        print(path)
    for i in g1[curr]:
        if(visited[i]):
            continue
        visited[i]=True
        parent[i]=curr
        distance[i]=distance[curr]+g1[curr][i]
        q.put((distance[i],i))